MeshLib/MeshGenerators/RasterToMesh.cpp

```diff
@@ -40,208 +44,194 @@ MeshLib::Mesh* RasterToMesh::convert(
     UseIntensityAs intensity_type,
     std::string const& array_name)
 {
-    if ((elem_type != MeshElemType::TRIANGLE) && (elem_type != MeshElemType::QUAD))
+    if ((elem_type != MeshElemType::TRIANGLE) &&
+        (elem_type != MeshElemType::QUAD) &&
+        (elem_type != MeshElemType::HEXAHEDRON) &&
+        (elem_type != MeshElemType::PRISM))
     {
-        ERR("VtkMeshConverter::convertImgToMesh(): Invalid Mesh Element Type.");
+        ERR("Invalid Mesh Element Type.");
         return nullptr;
     }
 
-    vtkSmartPointer<vtkDataArray> pixelData = vtkSmartPointer<vtkDataArray>(img->GetPointData()->GetScalars());
     int* dims = img->GetDimensions();
+    if (((elem_type == MeshElemType::TRIANGLE) || (elem_type == MeshElemType::QUAD)) &&
+        dims[2] != 1)
+    {
+        ERR("Triangle or Quad elements cannot be used to construct meshes from 3D rasters.");
+        return nullptr;
+    }
+
+    vtkSmartPointer<vtkDataArray> pixelData = vtkSmartPointer<vtkDataArray>(img->GetPointData()->GetScalars());
     int nTuple = pixelData->GetNumberOfComponents();
     if (nTuple < 1 || nTuple > 4)
     {
         ERR("VtkMeshConverter::convertImgToMesh(): Unsupported pixel composition!");
         return nullptr;
     }
 
-    MathLib::Point3d const orig (std::array<double,3>{{origin[0], origin[1], origin[2]}});
-    GeoLib::RasterHeader const header =
-        {static_cast<std::size_t>(dims[0]), static_cast<std::size_t>(dims[1]), orig, scalingFactor, -9999};
-    const std::size_t incHeight = header.n_rows+1;
-    const std::size_t incWidth  = header.n_cols+1;
-    std::vector<double> pix_val (incHeight * incWidth, std::numeric_limits<double>::max());
-    std::vector<bool> pix_vis (header.n_rows * header.n_cols, false);
+    MathLib::Point3d const orig(std::array<double, 3>{ {origin[0], origin[1], origin[2]}});
+    GeoLib::RasterHeader const header = {static_cast<std::size_t>(dims[0]),
+                                         static_cast<std::size_t>(dims[1]),
+                                         static_cast<std::size_t>(dims[2]),
+                                         orig,
+                                         scalingFactor,
+                                         -9999};
 
-    for (std::size_t i = 0; i < header.n_rows; i++)
-        for (std::size_t j = 0; j < header.n_cols; j++)
+    double *pix = new double[header.n_cols * header.n_rows * header.n_depth];
```

Comment:
Please use automatic memory management e.g. using std::vector, or unique_ptr.

Reply:
How can I access indices with a unique pointer? I.e. pix[index]?

Reply:
Yes, you [can](http://en.cppreference.com/w/cpp/memory/unique_ptr/operator_at). A `unique_ptr` is a pointer.
But I'd suggest to use `vector`.

Reply:
I'd prefer a double* because input software uses these (e.g. NetCDF or VTK). If I used a vector I would need to copy all the data, despite only needing it only once in the very beginning to get the initial data structure. It'd slow down the whole process.

```
std::unique_ptr<double> pix ( new double[header.n_cols * header.n_rows * header.n_depth] );
pix[5] = header.no_data;
```
doesn't work (at least in MSVS) and gives the following error msg:

```
error C2676: Binärer Operator "[": "std::unique_ptr<double,std::default_delete<_Ty>>" definiert diesen Operator oder eine Konvertierung in einen für den vordefinierten Operator geeigneten Typ nicht
2>        with
2>        [
2>            _Ty=double
2>        ]
```

Reply:
Using vector is much easier. I'm not sure where the copy is needed, but have a look on this change [in my repo](https://github.com/endJunction/ogs/commit/bde4153b7c3e5af028962dbefb3a528238c88bde), please.